src/playlist-loader.js

```diff
@@ -379,57 +389,104 @@
+    // 1) Walk backward until we find the first segment with timeline
     // information that is earlier than `time`
-    targetDuration = this.media_.targetDuration || 10;
-    i = this.media_.segments.length;
-    while (i--) {
+    for (i = lastSegment; i > 0; i--) {
       segment = this.media_.segments[i];
       if (segment.end !== undefined && segment.end <= time) {
-        time -= segment.end;
+        startIndex = i + 1;
+        knownStart = segment.end;
+        if (startIndex >= numSegments) {
+          // Somehow the last segment claims to end *before* the time we are
+          // searching for so just return it
+          return lastSegment;
+        }
         break;
       }
       if (segment.start !== undefined && segment.start < time) {
-
         if (segment.end !== undefined && segment.end > time) {
           // we've found the target segment exactly
           return i;
         }
+        startIndex = i;
+        knownStart = segment.start;
+        break;
+      }
+    }
 
-        time -= segment.start;
-        time -= segment.duration || targetDuration;
-        if (time < 0) {
-          // the segment with start information is also our best guess
-          // for the momment
+    // 2) Walk forward until we find the first segment with timeline
+    // information that is greater than `time`
+    for (i = 0; i < this.media_.segments.length; i++) {
+      segment = this.media_.segments[i];
+      if (segment.start !== undefined && segment.start > time) {
+        endIndex = i - 1;
+        knownEnd = segment.start;
+        if (endIndex < 0) {
+          // Somehow the first segment claims to start *after* the time we are
+          // searching for so just return it
+          return 0;
+        }
+        break;
+      }
+      if (segment.end !== undefined && segment.end > time) {
+        if (segment.start !== undefined && segment.start < time) {
+          // we've found the target segment exactly
           return i;
         }
+        endIndex = i;
+        knownEnd = segment.end;
         break;
       }
     }
-    i++;
 
-    // 2) Walk forward, testing each segment to see if `time` falls within it
-    for (j = i; j < this.media_.segments.length; j++) {
-      segment = this.media_.segments[j];
-      time -= segment.duration || targetDuration;
+    if (startIndex !== undefined) {
+      // We have a known-start point that is before our desired time so
+      // walk from that point forwards
+      time = time - knownStart;
+      for (i = startIndex; i < (endIndex || numSegments); i++) {
+        segment = this.media_.segments[i];
+        time -= segment.duration || targetDuration;
+        if (time < 0) {
+          return i;
+        }
+      }
 
-      if (time < 0) {
-        return j;
+      if (i === endIndex) {
+        // We haven't found a segment but we did hit a known end point
+        // so fallback to "Algorithm Jon" (ie. interpolation)
```

Comment:
"Algorithm Jon" is a funny name but I think we should be less entertaining, more informative in the comments.